cmd/release-controller/http.go

```diff
@@ -710,7 +710,48 @@ func (c *Controller) httpReleaseInfo(w http.ResponseWriter, req *http.Request) {
 
 	renderVerifyLinks(w, *info.Tag, info.Release)
 
-	if upgradesTo := c.graph.UpgradesTo(tag); len(upgradesTo) > 0 {
+	upgradesTo := c.graph.UpgradesTo(tag)
+
+	type releaseInfoStub struct {
+		Metadata   *struct {
+			Version  string   `json:"version"`
+			Previous []string `json:"previous"`
+		}               `json:"metadata"`
+	}
+	tagReleaseJson, err := c.releaseInfo.ReleaseInfo(tagPull)
+	var requiredUpgrades []string
+	var tagReleaseInfo releaseInfoStub
+	var missingUpgrades int
+	err = json.Unmarshal([]byte(tagReleaseJson), &tagReleaseInfo)
+	if err == nil && tagReleaseInfo.Metadata != nil {
+		requiredUpgrades = tagReleaseInfo.Metadata.Previous
+	}
+	upgradeFound := make(map[string]bool)
+	if len(requiredUpgrades) > 0 {
+		for _, u := range upgradesTo {
+			upgradeFound[u.From] = true
+		}
+		for _, from := range requiredUpgrades {
+			if !upgradeFound[from] {
+				 upgradesTo= append(upgradesTo, UpgradeHistory{
+					From:    from,
+					To:      tag,
+					Total:   -1,
+				})
+				if missingUpgrades == 0 {
+					fmt.Fprintf(w, `<span title="required upgrades listed in release info for which tests are missing"><p>Missing upgrades: </p><p>`)
```

Comment:
Don't print in the loop.  Create an array of upgrades that are missing and then render it in a loop afterwards.  Don't mix data operations and data rendering.